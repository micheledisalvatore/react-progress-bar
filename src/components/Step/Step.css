.step {
    flex-grow: 1;
    position: relative;
}

.step__dot {
    text-align: center;
    display: inline-block;
    position: absolute;
    top: -2.3em;
    left: 0;
}

.step__description {
    position: relative;
    letter-spacing: 1px;
    font-weight: bold;
    color: #aaa;
    top: -1em;
}

.step__description--is-active {
    color: #5131D7;
    transition: color 0s linear;
    transition-delay: 500ms;
}