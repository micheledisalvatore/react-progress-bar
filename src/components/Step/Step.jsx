import React from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';

import Connection from '../Connection';
import Dot from '../Dot';

import './Step.css';

export const Step = ({title, isFirst, isActive, isClickable, onClick, step}) => {
  const stepDescriptionClassName = classNames('step__description', {
    'step__description--is-active': isActive,
  });

  return (
    <div className="step">
      { !isFirst && <Connection isActive={isActive}/> }
      <div className={stepDescriptionClassName}>{title}</div>
      <Dot isActive={isActive} onClick={onClick} step={step} isClickable={isClickable} />
    </div>
  )
};

Step.propTypes = {
  isActive: PropTypes.bool,
  isFirst: PropTypes.bool,
  title: PropTypes.string.isRequired,
  onClick: PropTypes.func.isRequired,
  step: PropTypes.oneOf([1, 2, 3, 4, 5]).isRequired,
};

Step.defaultProps = {
  isActive: false,
  isFirst: false,
};