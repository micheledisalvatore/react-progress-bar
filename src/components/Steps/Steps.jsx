import React from 'react';
import PropTypes from 'prop-types';

import Step from '../Step';

import './Steps.css';

export const Steps = ({total, current, onClick}) => (
  <div className="steps">
    {[...(new Array(total))].map((e, i) => (
      <Step
        key={i}
        title={`Step ${i + 1}`}
        step={i + 1}
        isFirst={i === 0}
        isActive={i < current}
        onClick={onClick}
      />
    ))}
  </div>
);

Steps.propTypes = {
  total: PropTypes.oneOf([2,3,4,5]).isRequired,
  current: PropTypes.oneOf([1,2,3,4,5]).isRequired,
  onClick: PropTypes.func.isRequired,
};