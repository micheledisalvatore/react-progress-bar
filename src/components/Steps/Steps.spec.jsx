import React from 'react';
import { shallow } from 'enzyme';

import Step from '../Step';

import { Steps } from './Steps';

describe('Given a Steps component', () => {
  describe('when it is rendered', () => {
    const onClickMock = jest.fn();
    const requiredProps = {
      total: 2,
      current: 1,
      onClick: onClickMock,
    };
    const component = <Steps {...requiredProps} />;
    let wrapper;

    beforeEach(() => {
      wrapper = shallow(component);
    });

    afterEach(() => {
      onClickMock.mockReset();
    });

    it('should match the snapshot', () => {
      expect(wrapper.getElement()).toMatchSnapshot();
    });

    it('should render 2 Step component', () => {
      expect(wrapper.find(Step).length).toBe(2);
    });

    describe('and the first step', () => {
      let firstStep;

      beforeEach(() => {
        firstStep = wrapper.find(Step).at(0);
      });

      it('should have a title prop', () => {
        expect(firstStep).toHaveProp('title', 'Step 1');
      });

      it('should have a step prop', () => {
        expect(firstStep).toHaveProp('step', 1);
      });

      it('should have a isFirst prop to be true', () => {
        expect(firstStep).toHaveProp('isFirst', true);
      });

      it('should have a isActive prop to be true', () => {
        expect(firstStep).toHaveProp('isActive', true);
      });

      describe('and it is clicked', () => {
        beforeEach(() => {
          firstStep.simulate('click', 1);
        });

        it('should call onClick', () => {
          expect(onClickMock).toHaveBeenCalledWith(1);
        })
      })
    });

    describe('and the second step', () => {
      let secondStep;

      beforeEach(() => {
        secondStep = wrapper.find(Step).at(1);
      });

      it('should have a title prop', () => {
        expect(secondStep).toHaveProp('title', 'Step 2');
      });

      it('should have a step prop', () => {
        expect(secondStep).toHaveProp('step', 2);
      });

      it('should have a isFirst prop to be false', () => {
        expect(secondStep).toHaveProp('isFirst', false);
      });

      it('should have a isActive prop to be false', () => {
        expect(secondStep).toHaveProp('isActive', false);
      });

      describe('and it is clicked', () => {
        beforeEach(() => {
          secondStep.simulate('click', 2);
        });

        it('should call onClick', () => {
          expect(onClickMock).toHaveBeenCalledWith(2);
        })
      })
    })
  });
});