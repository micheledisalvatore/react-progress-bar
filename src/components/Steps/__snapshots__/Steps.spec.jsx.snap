// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a Steps component when it is rendered should match the snapshot 1`] = `
<div
  className="steps"
>
  <Unknown
    isActive={true}
    isFirst={true}
    onClick={[Function]}
    step={1}
    title="Step 1"
  />
  <Unknown
    isActive={false}
    isFirst={false}
    onClick={[Function]}
    step={2}
    title="Step 2"
  />
</div>
`;
